bastionzero/target/dbtarget/dbtarget.go

```diff
@@ -81,39 +103,247 @@ func makeDbTargetDataSourceSchema(opts *target.BaseTargetDataSourceAttributeOpti
+}
+func (t *dbTargetResourceModel) SetRemoteHost(value types.String) { t.RemoteHost = value }
+func (t *dbTargetResourceModel) SetRemotePort(value types.Int64)  { t.RemotePort = value }
+func (t *dbTargetResourceModel) SetLocalPort(value types.Int64)   { t.LocalPort = value }
+
+// setDbTargetResourceAttributes populates the TF schema data from a db target
+// API object.
+func setDbTargetResourceAttributes(ctx context.Context, schema *dbTargetResourceModel, dbTarget *targets.DatabaseTarget) {
+	target.SetBaseTargetAttributes(ctx, schema, dbTarget)
+	target.SetBaseVirtualTargetAttributes(ctx, schema, dbTarget)
+
+	schema.DatabaseAuthenticationConfig = FlattenDatabaseAuthenticationConfig(ctx, &dbTarget.DatabaseAuthenticationConfig)
+}
+
+func makeDbTargetResourceSchema(ctx context.Context) map[string]resource_schema.Attribute {
+	// Valid constants for fields in `database_authentication_config`. These are
+	// the only values the BastionZero backend currently accepts
+	validAuthenticationTypes := []string{
+		dbauthconfig.Default,
+		dbauthconfig.SplitCert,
+		dbauthconfig.ServiceAccountInjection,
+	}
+	validCloudServiceProviders := []string{
+		dbauthconfig.AWS,
+		dbauthconfig.GCP,
+	}
+	validDatabases := []string{
+		dbauthconfig.CockroachDB,
+		dbauthconfig.MicrosoftSQLServer,
+		dbauthconfig.MongoDB,
+		dbauthconfig.MySQL,
+		dbauthconfig.Postgres,
+	}
+
+	return map[string]resource_schema.Attribute{
+		"id": resource_schema.StringAttribute{
+			Computed:    true,
+			Description: "The target's unique ID.",
+			PlanModifiers: []planmodifier.String{
+				stringplanmodifier.UseStateForUnknown(),
+			},
+		},
+		"name": resource_schema.StringAttribute{
+			Required:    true,
+			Description: "The target's name.",
+			Validators: []validator.String{
+				stringvalidator.LengthAtLeast(1),
+			},
+		},
+		"type": resource_schema.StringAttribute{
+			Computed:    true,
+			Description: fmt.Sprintf("The target's type (constant value `%s`).", targettype.Db),
+			Default:     stringdefault.StaticString(string(targettype.Db)),
+			PlanModifiers: []planmodifier.String{
+				stringplanmodifier.UseStateForUnknown(),
+			},
+		},
+		"status": resource_schema.StringAttribute{
+			Computed:    true,
+			Description: fmt.Sprintf("The target's status %s.", internal.PrettyOneOf(targetstatus.TargetStatusValues())),
+			Validators: []validator.String{
+				stringvalidator.OneOf(bastionzero.ToStringSlice(targetstatus.TargetStatusValues())...),
+			},
+		},
+		"environment_id": resource_schema.StringAttribute{
+			Required:    true,
+			Description: "The target's environment's ID.",
+			Validators: []validator.String{
+				bzvalidator.ValidUUIDV4(),
+			},
+		},
+		"last_agent_update": resource_schema.StringAttribute{
+			Computed:    true,
+			Description: fmt.Sprintf("The time this target's proxy agent last had a transition change in status %s. Null if there has not been a single transition change.", internal.PrettyRFC3339Timestamp()),
+		},
```

Comment:
Actually one more question about this attribute of the resource. Since this value is constantly changing and shows up in the tf plan when other changes are made does it really make sense to include it in the resource? What advanatges are there for having it in the resource? Just to use as an output value? Same argument could be made for the `status`, `region` attributes as well

```
Terraform will perform the following actions:

  # bastionzero_db_target.example will be updated in-place
  ~ resource "bastionzero_db_target" "example" {
      ~ agent_public_key               = "n/a" -> (known after apply)
      ~ agent_version                  = "n/a" -> (known after apply)
        id                             = "3eb923b8-ccf0-4457-8f07-46a5becdac2b"
      + last_agent_update              = (known after apply)
        name                           = "tf-example-psql-db-target"
      - proxy_environment_id           = "312d0fc0-5679-4a47-9f5a-af4884d3bee5" -> null
      + proxy_target_id                = "b0c0a199-8e74-4a05-ac07-e904a33a0e6f"
      ~ region                         = "n/a" -> (known after apply)
      ~ status                         = "Online" -> (known after apply)
        # (5 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

Reply:
> Since this is just a "computed" value returned from the api and could be constantly changing and shows up in the tf plan when other changes are made

Just to reiterate what I think you're talking about above...It will only show `(known after apply)` like in the example you posted when one of the `bastionzero_db_target` resource's configurable attributes is modified. If you refer to one of the resource's read-only/computed attributes in an output or somewhere else (like another resource or data source) _without_ changing the resource's configurable attributes, it should show the refreshed value (if changed). Here is an example when I tried to restart the backing agent of a db target with some outputs that refer to `last_agent_update` and `status` (outputs start with `foo` below):

```
Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the last "terraform apply" which may have affected this plan:

  # bastionzero_db_target.example has changed
  ~ resource "bastionzero_db_target" "example" {
        id                             = "c7e9f2c5-e078-4699-94a5-f2ac20108b40"
      ~ last_agent_update              = "2023-11-15T19:29:04Z" -> "2023-11-15T19:34:28Z"
        name                           = "example-splitcert-db-target"
      ~ status                         = "Online" -> "Restarting"
        # (9 unchanged attributes hidden)
    }


Unless you have made equivalent changes to your configuration, or ignored the relevant attributes using ignore_changes, the following plan may include
actions to undo or respond to these changes.

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Changes to Outputs:
  ~ foo_status                      = "Online" -> "Restarting"
  ~ foo_update                      = "2023-11-15T19:29:04Z" -> "2023-11-15T19:34:28Z"

You can apply this plan to save these new output values to the Terraform state, without changing any real infrastructure.
```

> does it really make sense to include it in the resource?

All of these attributes: `status`,  `last_agent_update`, `region`, `agent_public_key`, etc. are part of the CRUD BastionZero API responses and Hashicorp recommends making the TF resource/attribute schema map the underlying API as close as possible:

> A Terraform resource and associated schema should follow the naming and structure of the API, unless the it degrades the user experience or works in a way counter to a user's expectations of Terraform.

Source: https://developer.hashicorp.com/terraform/plugin/best-practices/hashicorp-provider-design-principles#resource-and-attribute-schema-should-closely-match-the-underlying-api

That link includes the benefits for following this practice.

Furthermore, it would be a bit jarring for the `bastionzero_db_target` _data source_ to include these attributes, but _not_ in the resource itself. Meaning if a user wanted to refer to any of these fields, they would have to do an extra API call / use the data source for a resource they've already created in TF. But under the hood, the resource should already have these fields since the refresh/planning procedure calls the GET API already for the resource (and that API returns those fields anyways).

---

Also, if you're wondering why some of the other computed/read-only values do not display `(known after apply)`, this is because I've added plan modifiers that prevent this behavior for attributes that cannot change even after update--`id` and `type` are examples of this.

I agree though it can look a bit noisy. But this is also what happens in `bastionzero_environment.targets` as well (I think you had a similar offline comment about that a long time ago).

Lastly, in the example you mentioned above, all of those computed attributes _can_ potentially change after update, especially when going from `proxy_environment_id`  --> `proxy_target_id` (or changing `proxy_target_id` to another proxy target):

- `agent_public_key` will be different if `proxy_target_id` changes to a different underlying agent or if you switch from proxy env --> proxy target (n/a --> to actual key)
- `agent_version` same logic as above
- `last_agent_update` same logic as above
- `region` same logic as above
- `status` same logic as above (e.g. switching to a diff proxy target whose status is no longer `Online`)

Reply:
Thanks for the detailed explanation and links those were helpful. I definitely agree with you overall we should keep them in the resource.

> It will only show (known after apply) like in the example you posted when one of the bastionzero_db_target resource's configurable attributes is modified.

Yup true so that does makes this considerably less annoying since it wont ever show up as a update in the terraform plan unless there is a real configurable change.

> All of these attributes: status, last_agent_update, region, agent_public_key, etc. are part of the CRUD BastionZero API responses and Hashicorp recommends making the TF resource/attribute schema map the underlying API as close as possible:

Alright i guess that makes sense. I do see the value in that but was still wondering if making it a subset of the api would also make sense and reduce noise. I was struggling to think of any practical reasons why `last_agent_update` or  `status` would need to be referenced in terraform (other than as output values for display purposes maybe). Its a little easier to see how something like `agent_public_key` may need to be referenced in another resource. I guess technically you could have other resources depend on the `status` or `last_agent_update` attributes but then its going to always cascade into plan changes when those attributes change. Practically speaking i dont think anyone would want that. All that said i guess there really isnt much harm in including them just in case. 

> Furthermore, it would be a bit jarring for the bastionzero_db_target data source to include these attributes, but not in the resource itself. Meaning if a user wanted to refer to any of these fields, they would have to do an extra API call / use the data source for a resource they've already created in TF. But under the hood, the resource should already have these fields since the refresh/planning procedure calls the GET API already for the resource (and that API returns those fields anyways).

yeah point very well taken here - it would be annoying to have to use a data source to reference one of these attributes when you the resource is already part of your terraform model.  So if we have the attributes in the data source we should also have them in the resource.